### ISSUE TITLE
Parallel particle evaluation

### ISSUE THREAD

**Issue by soamaven** [ljvmiranda921/pyswarms#258]
**Is your feature request related to a problem? Please describe.**
I know I have an expensive function to evaluate, I am doing a simulation based optimization of a hyper parameter space. I am planning to use HPC environment because of this. Can particles be evaluated in parallel? Synchronously or even asynchronously? I thought an [approach similar to this](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1769316/) would be interesting (let me know if article is inaccessible).

**Describe the solution you'd like**
A switch to enable threads for particle evaluations. In particular, I am fine with firing off `os.system()` commands and waiting for return. 

**Describe alternatives you've considered**
I supposed I could write a sub-script that executes the swarm in parallel, evaluates the target function for each particle, and returns those values to the optimizer. This wouldn't allow for async optimization. Also, I could also try to write another optimizer, one that takes an option for the number of parallel particles.

What is the best way to go here? Still digging into this API currently.

**Comment by ljvmiranda921** [comment]
Hi @soamaven , 

Thank you for this issue. To be honest I've been meaning to add this feature [since last August](https://github.com/ljvmiranda921/pyswarms/wiki/Development-Roadmap), especially that I see a lot of use-cases involving very expensive objective functions. 

I would love to work on this, it's just that we're still finding time (especially we have a backlog of refactor issues to do) to dev (most of us have day jobs or at uni). Hopefully I can work on this on Q1 or Q2 next year, we just need to make a new release that solves the past issues.

If you can help us develop a module or base framework for this, then we'd really appreciate it! :smile:

**Comment by soamaven** [comment]
Thanks!

Thinking out loud a bit below, let me know where I am misunderstanding the code.:

So it looks like it would be easiest to just implement a custom optimization loop! So its probably easiest to write an optimization loop that parallelizes calculating the `numpy.ndarray` cost in parallel, either in an async or sync fashion, where in the async loop, it would be a nested iteration/async loop. 

I am sure there is demand for making parallelization transparent though. My thoughts on that are probably things you are aware of.

It looks like an inheritable and simple `multiprocessing.pool.map()` implementation within the optimizer base might be hard to do since the general `SwarmOptimizer` requires each child to implement `optimize`. It also actually looks like there is some code repetition between the different optimizers that could maybe be done away with by implementing those common portions of the algorithm (update positions, update bests, etc.) of the swarm loop higher in the hierarchy.

A true async approach would require a new optimizer and/or modifications to the current optimizers, with the async calls wrapping the iteration loop body. The finished particles would get their velocity updated as each particle finishes. (I don't know a good algorithm for choosing which particle should be evaluated next as there will presumably be a queue -- sequentially is the simplest though maybe not smartest way to do it.) Maybe an optimizer factory would be good here? Instantiating an optimizer would go through the factory and it would return the proper version of an optimizer (sequential, sync, or async) which could work for all of them. 

Also, is the algorithm reproducing work [here](https://github.com/ljvmiranda921/pyswarms/blob/b83e99be17d3b05a5948b73ad11fd2a0f6b0ce36/pyswarms/single/global_best.py#L165)? Why is the `pbest` cost not stored from a previous iteration so the objective function doesn't need to be re-evaluated?

**Comment by stevenbw** [comment]
@ljvmiranda921 I am more than happy to help out on this :)

**Comment by ljvmiranda921** [comment]
Hey @stevenbw ! Thanks a lot and welcome to PySwarms! We're planning to have this one open for the next version! Right now we're just doing our best to release a new version (just clean up some documentation, etc.) then once we're done, I'll ping you right away!

Wanna help me triage in this stuff? @whzup ? Will be back by Feb!

**Comment by stevenbw** [comment]
Thank you, do let me know if there is anything else I can help out with in the meantime :)

**Comment by starfriend10** [comment]
I just want to say that this feature will be extremely helpful because my case could cost for days based on a single PC! 
I tried tisimst's pyswarm 0.7 which has an argument of "processes", but it seems that it doesn't work for me in a 20-cores desktop.

**Comment by ljvmiranda921** [comment]
We'll try our best @starfriend10 , as of now most of us are really swamped by our day jobs and we haven't estimated the complexity of adding this feature (as for myself, I need to learn how subprocessing works). I guess the next critical feature for the 0.5.0 release would be to avoid computing the cost function twice per iteration, that I think we can do in the short-term.

cc: @whzup

**Comment by starfriend10** [comment]
@ljvmiranda921 Sure, I saw the discussion you had about the repeating computation problem, it's glad to hear that you will fix it in the near future. 
I don't know multiprocessing neither.
Although I don't know if it will be useful, I did a little customization (minor changes) of the code for my case, probably will share my experience in near future when I have more time.

**Comment by danielcorreia96** [comment]
Hey @ljvmiranda921, just wanted to say that I'm interested in working on this since the objective function I'm using for my project is fairly expensive and I'm happy to help.

From what I can tell, there are 3 main features being discussed in this issue:
1. **Parallel particle evaluation using **multiple threads****
   - This probably won't improve performance due to Python's GIL 🤔 
2. **Parallel particle evaluation using **multiple processes****
   - The multiprocessing library should be enough to do this (I've some experience with it)
3. ****Asynchronous** particle evaluation** (using master-slaves architecture)
   - I've read the paper and it looks interesting

Additionally, @soamaven made a good point regarding the repetitions in the optimize() function, which highlights some refactoring opportunities.

So, what do you think? Should we target the refactorings of the optimize() function first or just dive into each of these features? 😄

**Comment by ljvmiranda921** [comment]
Hi @danielcorreia96 ,


It would really be nice to have more hands on deck, especially that parallelism is (to be honest) not something I can definitely dive into! @whzup and I have been thinking of this feature for a while, and a lot of requests are coming in regarding the speed of our optimization algorithms. Having someone dedicated to this task will be very nice!

I think **Parallel particle evaluation using multiple processes** should be the easiest to implement. The design I imagine for this would be is to have a parameter called `n_jobs (int, default is -1 meaning use all available processing units)` that distributes the compute among that value. 

Also, as far as I remember the repetitions in the `optimize()` method should have already been solved by your #266 ? 

Sure, would you like to give this a try? 

cc: @whzup

**Comment by danielcorreia96** [comment]
@ljvmiranda921 I'll start working on it then 👍 

As far as design goes, we're mostly in agreement except that I think the **default value** should be `None`, as in "**don't paralelize the evaluation**", so that the user can control this part of the optimizer. Depending on the objective function and the execution context, **the overhead of paralelization may or may not be worth it**.

We can still have some kind of "magic number" that maps to "use all processing units" if you want though.

Also, regarding the repetitions, I think @soamaven was refering to the duplicate blocks of code in the `optimize` method across different classes.

For example, these two snippets from `global_best.py` and `local_best.py` highlight such duplications. You can easily find other situations like this in `optimize` method implementations (e.g. history update, position+velocity updates).
https://github.com/ljvmiranda921/pyswarms/blob/24a82e38c5f46cc4c414aa24de3d50c5fe5ad139/pyswarms/single/global_best.py#L199-L205
https://github.com/ljvmiranda921/pyswarms/blob/24a82e38c5f46cc4c414aa24de3d50c5fe5ad139/pyswarms/single/local_best.py#L227-L233

**Comment by whzup** [comment]
@ljvmiranda921 @danielcorreia96 May I provide some thoughts of mine to the design issue. I feel like we should decouple the parallel particle evaluation and the optimizers such that one would still have access to it when he builds his own optimizer. To be honest, I don't know if this is even possible but the optimizers should have as little of additional functionality compared to the backend as possible, shouldn't they?

**Comment by ljvmiranda921** [comment]
@whzup makes a lot of sense. 

Is it possible to create a pyswarms.backend.multiprocessingand call that whenever the user needs to implement multiproc when building his own optimizer?

This then means that our optimizers will be calling from the backend and the multiproc, the same way that logging (the Reporter module) is decoupled from the high level abstractions (optimizers) to the low-level ones (backend)

@danielcorreia96 yup, default is None, -1 for all CPUs, and any int will specify the number of CPUs. 

Will @whzup agree? Create a multiproc module similar in philosophy with the reporter module. 

Is this possible, @danielcorreia96?

**Comment by ljvmiranda921** [comment]
Not yet sure how pyswarms.backend.multiprocessing should be called, maybe

- as a context? 
- as a decorator?

Pardon the code block, currently on phone: 

from pyswarms.backend import multiproc

with multiproc(n_jobs=-1, **kwargs):
    # anything below here is executed in parallel


I think implementation wise, you can have the freedom to write as you please, @danielcorreia16  

I dont even think multiprocessing should be the module name (might clash with the standard lib), maybe something like parallel, or any related word haha

**Comment by soamaven** [comment]
> I dont even think multiprocessing should be the module name

concurrent? simultaneous?

> default is None, -1 for all CPUs, and any int will specify the number of CPUs

not difficult to catch an explicit keyword such as `"all"` as well, and protect/warn against more processes than CPUs?


Thanks for working on this everyone -- been wrapping up my thesis and am glad someone has taken the time. I like the strategy you all have created.

**Comment by whzup** [comment]
I found [deco](https://github.com/alex-sherman/deco) after a quick google search. It seems to be very easy but it would add to our dependencies (and we already have enough of those 😄)  and may restrict us in expanding on the multiprocessing feature.

I feel like a decorator might be a sensible solution (although I have absolutely no experience with contexts and what their use case is besides `with-open`). What do you guys think?

**Comment by danielcorreia96** [comment]
I'm not sure we should make a new module in backend for parallel evaluation **yet**. 

I'd argue we should start with a simpler implementation wherein we **abstract the objective function evaluation**.

Hence, instead of calling the evaluation function directly like this
`self.swarm.current_cost = objective_func(self.swarm.position, **kwargs)`

we would instead call an abstraction such as
`self.swarm.current_cost = compute_costs(self.swarm.position, objective_func, **kwargs)`

which would then **have an option for parallel execution**.

This `compute_costs` abstraction could be added to `operators` in a initial stage.

To clarify, my point is that we should **develop a simpler version first that only targets parallel objective function evaluations** and afterwards we can decide on how that should be refactored into a generic backend feature.

Also, I don't agree with using decorators to control/trigger parallelization since all of these "feature options" should be passed and kept inside the optimizer.

**Comment by whzup** [comment]
Hmm, I agree @danielcorreia96 maybe that's the best strategy at the moment 🤔.

**Comment by danielcorreia96** [comment]
Hey, so I've been playing around with some solutions for parallel objective function evaluation and I've reached a wall.

The main issue is that, currently, objective functions are expected to take an array of particles and return an array of values. In other words, **the objective function passed to the optimizer must evaluate all particles in a single call**.

This is troublesome because, in order to spread out the evaluations across multiple workers/cores, **we would like an objective function that only evaluates a single particle**.

To illustrate, consider the `sphere` function in `single_obj`.
```python
def sphere(x): # x: ndarray (n_particles, dimensions)
  j = (x ** 2.0).sum(axis=1)  
  return j     # j: ndarray (n_particles)
```
The current version receives an `ndarray` x (particles) and returns an `ndarray` j (objective function values). This makes it impossible to parallelize since all particles are evaluated in a single call.

An alternative, parallelizable, version would look like this:
```python
def alt_sphere(p):  # p: ndarray (dimensions,)
  j = (p ** 2.0).sum(axis=1)  
  return j          # j: numeric value
```

Since this version only evaluates one particle at a time, we can then distribute the computation using `multiprocessing` like this:

```python
import multiprocessing as mp
with mp.Pool(mp.cpu_count() - 1) as p:
  p.map(alt_sphere, self.swarm.position)
```

How should we proceed? 
- Should we upgrade pyswarms to support two types of objective functions (first and second version)? 
- Should we switch to the second version and move the aggregation of results into an `ndarray` someplace else?

From what I've seen, the predecessor of pyswarms (https://github.com/tisimst/pyswarm) considered the second version w.r.t to the objective function input/output contract. I'm not sure why we made a different design decision in pyswarms but I'm guessing there's a reason for that 🤔.

cc: @whzup @ljvmiranda921

**Comment by ljvmiranda921** [comment]
Hey @stevenbw ,

Thanks for your questions!

It's a design decision to vectorize swarm computations. In this case, evaluating an objective function across the swarm is "parallelized."  It's supposed to be a speedup compared to the previous pyswarm. I'd say we keep it and let it stay that way.

Still, parallelization can still be achieved with the current design by dividing the chuks of compute (say, when the objective function is heavy) across different workers. The computation is still vectorized, of course, but we can divide it on various chunks of data.

To answer your questions:

1. Nope, too clunky and I don't want to maintain two different ways on how objective function is written. The cost function decorator (#228) written by @whzup should already solve the transform from single-particle to all-particles.
2. Nope, although how would you propose to update the API? Without multiprocessing, we might resort to for-loops etc whatsoever. 

To proceed, think of this scenario:

- User wants a million particles in the swarm
- User has a costly objective function

For one iteration, 
- Running a fully-vectorised processed might overload a single CPU. 
- Doing a for-loop, similar to pyswarm, might require traversing all those particles (say `n_jobs=1`

What if we divide the swarm into chunks of 250k into four jobs?

Lastly, it is still possible to use the current API when evaluating a single-particle swarm, you just need to pass an ndarray of shape `(1,1)`. 

The goal for this feature then is to leverage both vectorization and parallelization capabilities. Try looking at [this blog post](https://datascience.blog.wzb.eu/2018/02/02/vectorization-and-parallelization-in-python-with-numpy-and-pandas/) for more information.

Thank you for this comment, @stevenbw . Wondering if you would like to help us maintain Pyswarms one day? Maybe after we have released this parallelization feature?

**Comment by stevenbw** [comment]
@danielcorreia96 - I think Lj is referring to you not me here 😄

**Comment by ljvmiranda921** [comment]
Whoops, my bad!! :laughing:

**Comment by danielcorreia96** [comment]
Hey @ljvmiranda921,

Thanks for the clarifications. 

I'll try to use a chunking approach so that we can have parallelization and vectorization at the same time 👍. 
I think that is still achievable with my initial proposal of a  "simple version" using an abstraction for the objective function call (see https://github.com/ljvmiranda921/pyswarms/issues/258#issuecomment-465617716)

Regarding the maintenance question, I'll be using pyswarms for my Msc thesis during 2019 so I'll be around to help out where I can 😄.

**Comment by ljvmiranda921** [comment]
Hi @danielcorreia96 ,

Cool! 
Yes, I think the abstraction you proposed should work. However, in case the codebase for parallelization became unexpectedly large, feel free to create your own `parallel` module :smile:

**Comment by danielcorreia96** [comment]
@ljvmiranda921 I've opened a draft PR where we can iterate on implementation-specific details of this feature, until we think it's good to go 😄

**Comment by ichbinjakes** [comment]
I have just used this library to try and find a good subset of features using the example in the docs [here](https://pyswarms.readthedocs.io/en/latest/examples/feature_subset_selection.html#feature-selection-and-the-objective-function). I changed the code slightly to execute in parallel:
```
from multiprocessing.pool import ThreadPool
def f(x):
    """Higher-level method to do classification in the
    whole swarm...
    """
    pool = ThreadPool(12)
    j = pool.map(particle_performance, x)
    return np.array(j)
```
Leaving the number of threads out of ThreadPool crashed my kernel in jupyter.
You could use joblib instead of ThreadPool to run the for loop that evaluates particle performance.

Thanks for all the effort put into this library.

**Comment by ljvmiranda921** [comment]
Closed by #312
